pkg/gitserver/controller/controller.go

```diff
@@ -0,0 +1,231 @@
+func bcEntry(bcName, ref string) string {
+	if len(ref) > 0 {
+		return fmt.Sprintf("%s.%s", bcName, ref)
+	}
+	return bcName
+}
+
+// HandleBuildConfigUpdate determines if a BuildConfig is using the local git server. If it is,
+// then store the BuildConfig's name in the associated repository
+func (c *BuildConfigController) HandleBuildConfigUpdate(bc *buildapi.BuildConfig) error {
+
+	// Get the repository name if the BuildConfig points to a repository on this git server
+	repoName, hasGitServerSource := c.gitServerRepository(bc)
+
+	if !hasGitServerSource {
+		// If the BuildConfig's source doesn't point to this git server,
+		// ensure that any references to it are removed from existing git repositories.
+		entry, inCache := c.BCCache[bc.Name]
+		if inCache {
+			if err := c.removeFromRepository(entry.Repository, bcEntry(bc.Name, entry.Reference)); err != nil {
+				return err
+			}
+			delete(c.BCCache, bc.Name)
+		}
+		return nil
+	}
+
+	removed := false
+	entry, inCache := c.BCCache[bc.Name]
+	if inCache && (entry.Repository != repoName || entry.Reference != bc.Spec.Source.Git.Ref) {
+		removed = true
+		// If already exists in cache, but either the repository or reference don't match anymore,
+		// remove the reference from the git repository
+		if err := c.removeFromRepository(entry.Repository, bcEntry(bc.Name, entry.Reference)); err != nil {
+			return err
+		}
+		delete(c.BCCache, bc.Name)
+	}
+
+	if !inCache || removed {
+		// If not already in the cache or it has been removed because of a mismatch, add to the
+		// repository config
+		if err := c.addToRepository(repoName, bcEntry(bc.Name, bc.Spec.Source.Git.Ref)); err != nil {
+			return err
+		}
+		c.BCCache[bc.Name] = BCCacheEntry{
+			Repository: repoName,
+			Reference:  bc.Spec.Source.Git.Ref,
+		}
+	}
+
+	return nil
+}
+
+func (c *BuildConfigController) HandleBuildConfigDelete(bc *buildapi.BuildConfig) error {
+	// First remove from the cache (and corresponding repository) if present
+	entry, inCache := c.BCCache[bc.Name]
+	if inCache {
+		if err := c.removeFromRepository(entry.Repository, bcEntry(bc.Name, entry.Reference)); err != nil {
+			return err
+		}
+		delete(c.BCCache, bc.Name)
+	}
+	// Ensure that even if it's not in the cache, that its
+	// corresponding repository is updated if there is one.
+	repoName, hasGitServerSource := c.gitServerRepository(bc)
+	if hasGitServerSource {
+		if !inCache || entry.Repository != repoName || entry.Reference != bc.Spec.Source.Git.Ref {
+			if err := c.removeFromRepository(repoName, bcEntry(bc.Name, bc.Spec.Source.Git.Ref)); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c *BuildConfigController) gitServerRepository(bc *buildapi.BuildConfig) (name string, isRepo bool) {
+	if bc.Spec.Source.Git == nil {
+		return "", false
+	}
+	parsedURL, err := url.Parse(bc.Spec.Source.Git.URI)
+	if err != nil {
+		return "", false
+	}
+	if parsedURL.Host != c.PublicURL.Host ||
```

Comment:
The fragility of knowing whether a build config points at you is the reason why I got blocked on the gitserver auto linking - because the git url could be any one of a hundred variations that doesn't quite match.  The simplest misconfiguration would totally break this logic silently, which means hard to debug issues for end users.

I think you need a stronger signal here.  I think there are three flows that are relevant:

1. User sets up gitserver, pushes, and hopes it works.  I would *expect* pushing to trigger the build with the same name as the repo via binary build.  That's straight up simpler than this flow.
2. User does the more complex "point build config at local git server because they don't have their own repo" - that's a detection flow, but when you detect you should record a stronger signal than just "the current URL happens to match", like an annotation
3. User wants to have their build config point at a remote, but mirror and do polling.  In this case you're going to need more than this check.